HSEARCH-4725 @ObjectProjection filtering (includeDepth, includePaths, excludePaths) to apply projection constructors down to a certain depth

yrodiere opened the pull request:
https://hibernate.atlassian.net/browse/HSEARCH-4725

@marko-bekhta This will most likely conflict with the changes you're working on for https://hibernate.atlassian.net/browse/HSEARCH-4876 . Feel free to wait until your PR for HSEARCH-4876 has been merged before you review this; I'll rebase and handle conflicts.

marko-bekhta commented on mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/search/definition/binding/impl/ProjectionConstructorBinder.java:
```diff
@@ -84,36 +83,16 @@ public PojoConstructorProjectionDefinition<T> bind() {
-
-	private ProjectionConstructorPath createPath(ProjectionConstructorBinder<?> firstElement, ProjectionConstructorPath child, int childPosition) {
-		ProjectionConstructorPath pathFromSelf = new ProjectionConstructorPath( new PojoConstructorIdentifier( constructor ), child, childPosition );
+	ProjectionConstructorPath createPath(ProjectionConstructorBinder<?> firstElement,
```
isn't used anymore ?

marko-bekhta commented on documentation/src/main/asciidoc/public/reference/_mapping-projection.adoc:
```diff
@@ -142,6 +142,7 @@ Constructor parameters can be annotated with explicit projection annotations suc
 For projections that would normally be <<mapping-projection-inner-inference,inferred automatically>>,
 this allows further customization, for example in a <<search-dsl-projection-field-mapping,field projection>>
 to set the target field path explicitly or to disable value conversion.
+This also allows <<search-dsl-projection-object-mapping-filters,breaking cycles of nested object projections>>.
```
maybe something like:
```suggestion
Or for <<search-dsl-projection-object-mapping,object projections>> this also allows <<search-dsl-projection-object-mapping-filters,breaking cycles of nested object projections>>.
```
since we don't mention object projection it feels like you can break a cycle with a field projection 😃

marko-bekhta commented on documentation/src/main/asciidoc/public/reference/_search-dsl-projection.adoc:
```diff
@@ -1143,6 +1149,190 @@ include::{sourcedirJava17}/org/hibernate/search/documentation/search/projection/
+
+This will work just fine for simpler use cases, but may lead to problems for more complex models:
+
+* If the projection constructor of the projected type declares many nested projection,
```
```suggestion
* If the projection constructor of the projected type declares many nested projections,
```

marko-bekhta reviewed (commented):
So `+3,401 −1,461` turned out not as scary as it seemed 😃.
I've left a few comments here and there but it looks nice! thanks 😃

yrodiere commented on mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/mapping/definition/programmatic/impl/PropertyMappingIndexedEmbeddedStepImpl.java:
```diff
@@ -66,7 +67,8 @@ public void contributeIndexMapping(PojoIndexMappingCollectorPropertyNode collect
 		collector.value( extractorPath ).indexedEmbedded(
-				definingTypeModel, actualPrefix, structure, includeDepth, includePaths, excludePaths,
+				definingType, actualPrefix, structure,
+				new TreeFilterDefinition( includeDepth, includePaths, excludePaths ),
```
We probably could, but I don't know if it really bring anything performance-wise. Maybe let's do that if one day we do performance tests and those tests show filter definitions as a major problem?

To be clear, I only added `TreeFilterDefinition.includeAll()` to provide a clear and easy way to create such filter for people who just don't care about filtering.

yrodiere commented:
Thanks! I addressed your comments and force-pushed.

> So `+3,401 −1,461` turned out not as scary as it seemed

Yes, turns out I was able to reuse most of the filtering we already had for `@IndexedEmbedded`. Which means that the implementation is much simpler, but also that I needed very little tests.

marko-bekhta commented:
Merging 🎉
